substrate/primitives/core/src/address_uri.rs

```diff
@@ -0,0 +1,258 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Little util for parsing an address URI. Replaces regular expressions.
+
+#[derive(Debug, PartialEq)]
+/// A container for results of parsing the address uri string.
+pub struct AddressUri<'a> {
+	pub ss58: Option<&'a str>,
+	pub paths: Vec<&'a str>,
+	pub pass: Option<&'a str>,
+}
+
+impl<'a> AddressUri<'a> {
+	fn extract_prefix(input: &mut &'a str, is_allowed: &dyn Fn(char) -> bool) -> Option<&'a str> {
+		let output = input.trim_start_matches(is_allowed);
+		let prefix_len = input.len() - output.len();
+		let prefix = if prefix_len > 0 { Some(&input[..prefix_len]) } else { None };
```

Comment:
This also means that we are expecting that there is at least a new line at the end of the URI, that sounds wrong.

Reply:
Sorry did not get your comment. This is has nothing to do with new line.
If provided `input` does not start with  `is_allowed` characters, then extracted prefix should be `None`. 
In this case the `output` will be equal to `input` and thus `prefix_len` will be 0.

Reply:
Sorry, I didn't checked how `trim_start_matches` works!